// ews-javascript-api/lib/TimeZoneInfo.js
"use strict";
var ArgumentException_1 = require("../js/Exceptions/ArgumentException");
var TimeSpan_1 = require("./TimeSpan");
var DateTime_1 = require("./DateTime");
var ExtensionMethods_1 = require("../js/ExtensionMethods");
var DayOfWeek_1 = require("../js/Enumerations/DayOfWeek");
/**
* TimeZoneInfo
*/
var TimeZoneInfo = (function () {
    function TimeZoneInfo() {
    }
    Object.defineProperty(TimeZoneInfo.prototype, "Id", {
        get: function () { return this._id; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimeZoneInfo.prototype, "DisplayName", {
        get: function () { return this._displayName || ExtensionMethods_1.StringHelper.Empty; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimeZoneInfo.prototype, "StandardName", {
        get: function () { return this._standardDisplayName || ExtensionMethods_1.StringHelper.Empty; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimeZoneInfo.prototype, "DaylightName", {
        get: function () { return this._daylightDisplayName || ExtensionMethods_1.StringHelper.Empty; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimeZoneInfo.prototype, "BaseUtcOffset", {
        get: function () { return this._baseUtcOffset; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimeZoneInfo.prototype, "SupportsDaylightSavingTime", {
        get: function () { return this._supportsDaylightSavingTime; },
        enumerable: true,
        configurable: true
    });
    return TimeZoneInfo;
}());
// constants for TimeZoneInfo.Local and TimeZoneInfo.Utc
TimeZoneInfo.UtcId = "UTC";
TimeZoneInfo.LocalId = "Local";
TimeZoneInfo.s_utcTimeZone = TimeZoneInfo.CreateCustomTimeZone(TimeZoneInfo.UtcId, TimeSpan_1.TimeSpan.Zero, TimeZoneInfo.UtcId, TimeZoneInfo.UtcId);
// used by GetUtcOffsetFromUtc (DateTime.Now, DateTime.ToLocalTime) for max/min whole-day range checks
TimeZoneInfo.s_maxDateOnly = new DateTime_1.DateTime(9999, 12, 31);
TimeZoneInfo.s_minDateOnly = new DateTime_1.DateTime(1, 1, 2);
exports.TimeZoneInfo = TimeZoneInfo;
(function (TimeZoneInfo) {
    var TimeZoneInfoOptions;
    (function (TimeZoneInfoOptions) {
        TimeZoneInfoOptions[TimeZoneInfoOptions["None"] = 1] = "None";
        TimeZoneInfoOptions[TimeZoneInfoOptions["NoThrowOnInvalidTime"] = 2] = "NoThrowOnInvalidTime";
    })(TimeZoneInfoOptions = TimeZoneInfo.TimeZoneInfoOptions || (TimeZoneInfo.TimeZoneInfoOptions = {}));
    var TimeZoneInfoResult;
    (function (TimeZoneInfoResult) {
        TimeZoneInfoResult[TimeZoneInfoResult["Success"] = 0] = "Success";
        TimeZoneInfoResult[TimeZoneInfoResult["TimeZoneNotFoundException"] = 1] = "TimeZoneNotFoundException";
        TimeZoneInfoResult[TimeZoneInfoResult["InvalidTimeZoneException"] = 2] = "InvalidTimeZoneException";
        TimeZoneInfoResult[TimeZoneInfoResult["SecurityException"] = 3] = "SecurityException";
    })(TimeZoneInfoResult || (TimeZoneInfoResult = {}));
    ;
    var AdjustmentRule = (function () {
        function AdjustmentRule(dateStart, dateEnd, daylightDelta, daylightTransitionStart, daylightTransitionEnd, baseUtcOffsetDelta, noDaylightTransitions) {
            AdjustmentRule.ValidateAdjustmentRule(dateStart, dateEnd, daylightDelta, daylightTransitionStart, daylightTransitionEnd, noDaylightTransitions);
            this._dateStart = dateStart;
            this._dateEnd = dateEnd;
            this._daylightDelta = daylightDelta;
            this._daylightTransitionStart = daylightTransitionStart;
            this._daylightTransitionEnd = daylightTransitionEnd;
            this._baseUtcOffsetDelta = baseUtcOffsetDelta;
            this._noDaylightTransitions = noDaylightTransitions;
        }
        Object.defineProperty(AdjustmentRule.prototype, "DateStart", {
            get: function () {
                return this._dateStart;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "DateEnd", {
            get: function () {
                return this._dateEnd;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "DaylightDelta", {
            get: function () {
                return this._daylightDelta;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "DaylightTransitionStart", {
            get: function () {
                return this._daylightTransitionStart;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "DaylightTransitionEnd", {
            get: function () {
                return this._daylightTransitionEnd;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "BaseUtcOffsetDelta", {
            /** @internal */
            get: function () {
                return this._baseUtcOffsetDelta;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "NoDaylightTransitions", {
            /** @internal */
            get: function () {
                return this._noDaylightTransitions;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AdjustmentRule.prototype, "HasDaylightSaving", {
            /** @internal */
            get: function () {
                return ((this.DaylightDelta != TimeSpan_1.TimeSpan.Zero)
                    || ((this.DaylightTransitionStart.TimeOfDay != DateTime_1.DateTime.MinValue)
                        || (this.DaylightTransitionEnd.TimeOfDay != DateTime_1.DateTime.MinValue.AddMilliseconds(1))));
            },
            enumerable: true,
            configurable: true
        });
        // IEquatable<AdjustmentRule>
        AdjustmentRule.prototype.Equals = function (other) {
            var equals = ((other != null)
                && ((this._dateStart == other._dateStart)
                    && ((this._dateEnd == other._dateEnd)
                        && ((this._daylightDelta == other._daylightDelta)
                            && (this._baseUtcOffsetDelta == other._baseUtcOffsetDelta)))));
            equals = (equals
                && (this._daylightTransitionEnd.Equals(other._daylightTransitionEnd) && this._daylightTransitionStart.Equals(other._daylightTransitionStart)));
            return equals;
        };
        AdjustmentRule.prototype.GetHashCode = function () {
            return this._dateStart.GetHashCode();
        };
        AdjustmentRule.CreateAdjustmentRule = function (dateStart, dateEnd, daylightDelta, daylightTransitionStart, daylightTransitionEnd, baseUtcOffsetDelta, noDaylightTransitions) {
            if (baseUtcOffsetDelta === void 0) { baseUtcOffsetDelta = TimeSpan_1.TimeSpan.Zero; }
            if (noDaylightTransitions === void 0) { noDaylightTransitions = false; }
            return new AdjustmentRule(dateStart, dateEnd, daylightDelta, daylightTransitionStart, daylightTransitionEnd, baseUtcOffsetDelta, noDaylightTransitions);
        };
        // ----- SECTION: internal utility methods ----------------*
        //
        // When Windows sets the daylight transition start Jan 1st at 12:00 AM, it means the year starts with the daylight saving on. 
        // We have to special case this value and not adjust it when checking if any date is in the daylight saving period. 
        //
        /** @internal */
        AdjustmentRule.prototype.IsStartDateMarkerForBeginningOfYear = function () {
            return !this.NoDaylightTransitions &&
                this.DaylightTransitionStart.Month == 1 && this.DaylightTransitionStart.Day == 1 && this.DaylightTransitionStart.TimeOfDay.Hour == 0 &&
                this.DaylightTransitionStart.TimeOfDay.Minute == 0 && this.DaylightTransitionStart.TimeOfDay.Second == 0 &&
                this._dateStart.Year == this._dateEnd.Year;
        };
        //
        // When Windows sets the daylight transition end Jan 1st at 12:00 AM, it means the year ends with the daylight saving on. 
        // We have to special case this value and not adjust it when checking if any date is in the daylight saving period. 
        //
        /** @internal */
        AdjustmentRule.prototype.IsEndDateMarkerForEndOfYear = function () {
            return !this.NoDaylightTransitions &&
                this.DaylightTransitionEnd.Month == 1 && this.DaylightTransitionEnd.Day == 1 && this.DaylightTransitionEnd.TimeOfDay.Hour == 0 &&
                this.DaylightTransitionEnd.TimeOfDay.Minute == 0 && this.DaylightTransitionEnd.TimeOfDay.Second == 0 &&
                this._dateStart.Year == this._dateEnd.Year;
        };
        //
        // ValidateAdjustmentRule -
        //
        // Helper function that performs all of the validation checks for the 
        // factory methods and deserialization callback
        //
        AdjustmentRule.ValidateAdjustmentRule = function (dateStart, dateEnd, daylightDelta, daylightTransitionStart, daylightTransitionEnd, noDaylightTransitions) {
            if (dateStart.Kind != DateTime_1.DateTimeKind.Unspecified && dateStart.Kind != DateTime_1.DateTimeKind.Utc) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeKindMustBeUnspecifiedOrUtc", "dateStart");
            }
            if (dateEnd.Kind != DateTime_1.DateTimeKind.Unspecified && dateEnd.Kind != DateTime_1.DateTimeKind.Utc) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeKindMustBeUnspecifiedOrUtc", "dateEnd");
            }
            if (daylightTransitionStart.Equals(daylightTransitionEnd) && !noDaylightTransitions) {
                throw new ArgumentException_1.ArgumentException("Argument_TransitionTimesAreIdentical", "daylightTransitionEnd");
            }
            if (dateStart > dateEnd) {
                throw new ArgumentException_1.ArgumentException("Argument_OutOfOrderDateTimes", "dateStart");
            }
            // This cannot use UtcOffsetOutOfRange to account for the scenario where Samoa moved across the International Date Line,
            // which caused their current BaseUtcOffset to be +13. But on the other side of the line it was UTC-11 (+1 for daylight).
            // So when trying to describe DaylightDeltas for those times, the DaylightDelta needs
            // to be -23 (what it takes to go from UTC+13 to UTC-10)
            if (daylightDelta.TotalHours < -23.0 || daylightDelta.TotalHours > 14.0) {
                throw new ArgumentException_1.ArgumentOutOfRangeException("daylightDelta", daylightDelta, "ArgumentOutOfRange_UtcOffset");
            }
            if (daylightDelta.TotalMilliseconds % TimeSpan_1.TimeSpan.MillisPerMinute != 0) {
                throw new ArgumentException_1.ArgumentException("Argument_TimeSpanHasSeconds", "daylightDelta");
            }
            if (dateStart != DateTime_1.DateTime.MinValue && dateStart.Kind == DateTime_1.DateTimeKind.Unspecified && dateStart.TimeOfDay != TimeSpan_1.TimeSpan.Zero) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeHasTimeOfDay", "dateStart");
            }
            if (dateEnd != DateTime_1.DateTime.MaxValue && dateEnd.Kind == DateTime_1.DateTimeKind.Unspecified && dateEnd.TimeOfDay != TimeSpan_1.TimeSpan.Zero) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeHasTimeOfDay", "dateEnd");
            }
        };
        return AdjustmentRule;
    }());
    TimeZoneInfo.AdjustmentRule = AdjustmentRule;
    var TransitionTime = (function () {
        function TransitionTime(timeOfDay, month, week, day, dayOfWeek, isFixedDateRule) {
            TransitionTime.ValidateTransitionTime(timeOfDay, month, week, day, dayOfWeek);
            this._timeOfDay = timeOfDay;
            this._month = month;
            this._week = week;
            this._day = day;
            this._dayOfWeek = dayOfWeek;
            this._isFixedDateRule = isFixedDateRule;
        }
        Object.defineProperty(TransitionTime.prototype, "TimeOfDay", {
            get: function () {
                return this._timeOfDay;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TransitionTime.prototype, "Month", {
            get: function () {
                return this._month;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TransitionTime.prototype, "Week", {
            get: function () {
                return this._week;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TransitionTime.prototype, "Day", {
            get: function () {
                return this._day;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TransitionTime.prototype, "DayOfWeek", {
            get: function () {
                return this._dayOfWeek;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TransitionTime.prototype, "IsFixedDateRule", {
            get: function () {
                return this._isFixedDateRule;
            },
            enumerable: true,
            configurable: true
        });
        // [Pure]
        // public override bool Equals(object obj) =>
        // obj is TransitionTime && Equals((TransitionTime)obj);
        // public static bool operator == (TransitionTime t1, TransitionTime t2) => t1.Equals(t2);
        // public static bool operator != (TransitionTime t1, TransitionTime t2) => !t1.Equals(t2);
        TransitionTime.prototype.Equals = function (other) {
            return;
            this._isFixedDateRule == other._isFixedDateRule &&
                this._timeOfDay == other._timeOfDay &&
                this._month == other._month &&
                (other._isFixedDateRule ?
                    this._day == other._day :
                    this._week == other._week &&
                        this._dayOfWeek == other._dayOfWeek);
        };
        TransitionTime.prototype.GetHashCode = function () {
            return (this._month ^ this._week << 8);
        };
        ;
        TransitionTime.CreateFixedDateRule = function (timeOfDay, month, day) {
            return new TransitionTime(timeOfDay, month, 1, day, DayOfWeek_1.DayOfWeek.Sunday, true);
        };
        TransitionTime.CreateFloatingDateRule = function (timeOfDay, month, week, dayOfWeek) {
            return new TransitionTime(timeOfDay, month, week, 1, dayOfWeek, false);
        };
        /// <summary>
        /// Helper function that validates a TransitionTime instance.
        /// </summary>
        TransitionTime.ValidateTransitionTime = function (timeOfDay, month, week, day, dayOfWeek) {
            if (timeOfDay.Kind != DateTime_1.DateTimeKind.Unspecified) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeKindMustBeUnspecified", "timeOfDay");
            }
            // Month range 1-12
            if (month < 1 || month > 12) {
                throw new ArgumentException_1.ArgumentOutOfRangeException("month", "ArgumentOutOfRange_MonthParam");
            }
            // Day range 1-31
            if (day < 1 || day > 31) {
                throw new ArgumentException_1.ArgumentOutOfRangeException("day", "ArgumentOutOfRange_DayParam");
            }
            // Week range 1-5
            if (week < 1 || week > 5) {
                throw new ArgumentException_1.ArgumentOutOfRangeException("week", "ArgumentOutOfRange_Week");
            }
            // DayOfWeek range 0-6
            if (dayOfWeek < 0 || dayOfWeek > 6) {
                throw new ArgumentException_1.ArgumentOutOfRangeException("dayOfWeek", "ArgumentOutOfRange_DayOfWeek");
            }
            if (timeOfDay.Year != 1 || timeOfDay.Month != 1 || timeOfDay.Day != 1 || (timeOfDay.TotalMilliSeconds % 1 != 0)) {
                throw new ArgumentException_1.ArgumentException("Argument_DateTimeHasTicks", "timeOfDay");
            }
        };
        return TransitionTime;
    }());
    TimeZoneInfo.TransitionTime = TransitionTime;
})(TimeZoneInfo = exports.TimeZoneInfo || (exports.TimeZoneInfo = {}));
exports.TimeZoneInfo = TimeZoneInfo;
